core/src/kura.rs

```diff
@@ -4,258 +4,504 @@
+            match block_reciever_guard.try_recv() {
+                Err(_) => {
+                    std::thread::sleep(std::time::Duration::from_millis(1));
+                    continue;
+                }
+                Ok(new_block) => {
+                    #[cfg(feature = "telemetry")]
+                    if new_block.header().height == 1 {
+                        iroha_logger::telemetry!(msg = iroha_telemetry::msg::SYSTEM_CONNECTED, genesis_hash = %new_block.hash());
+                    }
+
+                    let block_hash = new_block.hash();
+                    let serialized_block: Vec<u8> = new_block.encode_versioned();
+
+                    match strong_kura_reference
+                        .block_store
+                        .lock()
+                        .expect("lock on block store")
+                        .append_block_to_chain(&serialized_block)
+                    {
+                        Ok(()) => {
+                            strong_kura_reference
+                                .block_hash_array
+                                .lock()
+                                .expect("lock on block hash array")
+                                .push(block_hash);
+                            strong_kura_reference.broker.issue_send_sync(&ContinueSync);
+                        }
+                        Err(error) => {
+                            error!(%error, "Failed to write block. Kura will now init again.");
+                            panic!("It is unclear what we should do here.");
+                        }
+                    }
                 }
-                self.wsv.init(blocks).await;
-                let last_block = self.wsv.latest_block_hash();
-                let height = self.wsv.height();
-                self.broker
-                    .issue_send(sumeragi::message::Init { last_block, height })
-                    .await;
-            }
-            Err(error) => {
-                error!(%error, "Kura initialization failed. Try removing the peer's `blocks` directory and restarting the peer. You can also try rebuilding the peer altogether.");
-                panic!("Kura initialization failed");
             }
         }
     }
-}
 
-#[async_trait::async_trait]
-impl<IO: DiskIO> Handler<GetBlockHash> for KuraWithIO<IO> {
-    type Result = Option<HashOf<VersionedCommittedBlock>>;
-    async fn handle(&mut self, GetBlockHash { height }: GetBlockHash) -> Self::Result {
-        if height == 0 {
+    /// Get the hash of the block at the provided height.
+    #[allow(clippy::unwrap_in_result, clippy::expect_used)]
+    pub fn get_block_hash(&self, block_height: u64) -> Option<HashOf<VersionedCommittedBlock>> {
+        let hash_array_guard = self.block_hash_array.lock().expect("access hash array");
+        if block_height == 0 || block_height > hash_array_guard.len() as u64 {
```

Comment:
```suggestion
        if block_height == 0 || block_height > hash_array_guard.len().try_into().expect("`usize` should always fit in `u64`") {
```

Reply:
I think this is pointless. There are no 128-bit platforms. This cast cannot fail, ever. So sacrificing readability for "correctness" that isn't more correct in the real world isn't productive.